app/human-detector-app/src/auth/tokenManager.ts

```diff
@@ -0,0 +1,61 @@
+import { DiscoveryDocument, TokenResponse } from 'expo-auth-session';
+import User from '../../classes/User';
+import { getUserFromIDToken } from './helpers';
+
+/**
+ * Handles access to and refreshing of credentials for the backend.
+ */
+export default class TokenManager {
+  private tokenResponse: TokenResponse;
+
+  private discovery: DiscoveryDocument;
+
+  private clientId: string;
+
+  private constructor(
+    tokenResponse: TokenResponse,
+    discovery: DiscoveryDocument,
+    clientId: string
+  ) {
+    this.tokenResponse = tokenResponse;
+    this.discovery = discovery;
+    this.clientId = clientId;
+  }
+
+  /**
+   * Construct a TokenManager from an AuthSession token response, assuming that the response has
+   * all of the required fields (refresh token, expiration, and ID token).
+   * @param response AuthSession TokenResponse
+   * @param discovery endpoint discovery document
+   * @param clientId
+   * @returns a new TokenManager
+   */
+  public static fromTokenResponse(
+    response: TokenResponse,
+    discovery: DiscoveryDocument,
+    clientId: string
+  ) {
+    return new this(response, discovery, clientId);
+  }
+
+  /**
+   * Get the access token, refreshing as needed.
+   */
+  public async getAccessToken(): Promise<string> {
```

Comment:
How exactly did you test that this was working?  When I was trying to work with shouldRefresh() I was having issues with it always refreshing no matter what.

Reply:
I assumed it wasn't refreshing every time, but I just tested and it looks like it is. I dug a little bit and it looks like AuthSession has a refresh margin margin of 10 minutes by default ([source](https://github.com/expo/expo/blob/fe3bd0dc11905e694d2c3ad492e127135c8f2157/packages/expo-auth-session/src/TokenRequest.ts#L41-L43)), but our tokens are configured to expire after 5 minutes. I could change the code, or we could just bump the lifetime of our tokens up to something like 30 minutes.

This should be the only time we have to refresh; our refresh tokens (currently) don't expire, so we can refresh whenever we want to. The backend service calls `getAccessToken()` before each request, so I figured that would be a good time to refresh our tokens if we need to.

Reply:
Would bumping it up to 30 minutes be against like some standard?  If that means we don't have to refresh every time I think it makes more sense to me.  What do you think?

Reply:
There doesn't seem to be a standard, based on [the spec](https://www.rfc-editor.org/rfc/rfc6749#section-4.2.2) I think we're good as long as our tokens expire. I wasn't able to find anything helpful just Googling around either. I think a bump from 5 minutes to 30 minutes is reasonable, otherwise we would be constantly refreshing our tokens during regular usage

Reply:
That's fine by me if you're ok with making that change

Reply:
Sounds good. Bumped the expiration time up to 20 minutes and manually tested login with some logging thrown in, it no longer refreshes every time